File: src/main/java/com/example/AppConfig.java
package com.example;

import com.example.discount.DiscountPolicy;
import com.example.discount.RateDiscountPolicy;
import com.example.member.repository.MemberRepository;
import com.example.member.repository.MemoryMemberRepository;
import com.example.member.service.MemberService;
import com.example.member.service.MemberServiceImpl;
import com.example.order.service.OrderService;
import com.example.order.service.OrderServiceImpl;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class AppConfig {
	
	@Bean
	public MemberRepository memberRepository() {
		return new MemoryMemberRepository();
	}
	
	@Bean
	public MemberService memberService() {
		return new MemberServiceImpl(memberRepository());
	}
	
	@Bean
	public OrderService orderService() {
		return new OrderServiceImpl(memberRepository(), discountPolicy());
	}
	
	@Bean
	public DiscountPolicy discountPolicy() {
		return new RateDiscountPolicy();
	}
}

File: spring_core_principle_basic/src/main/kotlin/hello/core/AppConfig.kt
package hello.core

import hello.core.discount.DiscountPolicy
import hello.core.discount.RateDiscountPolicy
import hello.core.member.repository.MemberRepository
import hello.core.member.repository.MemoryMemberRepository
import hello.core.member.service.MemberService
import hello.core.member.service.MemberServiceImpl
import hello.core.order.OrderService
import hello.core.order.OrderServiceImpl


class AppConfig {
    fun memberService(): MemberService = MemberServiceImpl(memberRepository())

    fun orderService(): OrderService = OrderServiceImpl(memberRepository(), discountPolicy())

    private fun memberRepository(): MemberRepository = MemoryMemberRepository()

    private fun discountPolicy(): DiscountPolicy = RateDiscountPolicy()
}